### DISCUSSION TITLE
Option to change configuration from /dev/tty to stdout

### DISCUSSION THREAD

**Discussion by chenhayat** [crossbario/crossbar#2110] [Q&A]
Hi, I am trying to run `crossbar status` via popen process but it seems that because it's written to /dev/tty it cannot redirect the output and therefore I cannot read the status in python. Is there an option to change the log to write to stdout ?

**Comment by oberstet** [comment]
yeah, that might be the case, I seem to remember we've added special behavior .. in the context of multi-process (like Crossbar.io) Docker containers stuff gets even more interesting;)

anyways, in general it would be good if you'd posted more: the complete output of `crossbar status` when run at the interactive terminal of the Crossbar.io you do run when deploying (!) like e.g. from within a Docker container. for a start, this will allow the precise versions of code you actually run - _unless_ you've modified that code and run from there. stuff like that.

**Reply by chenhayat** [reply]
Thanks for your reply :)
Crossbar version is v21.3.1.
Docker is no used in this case, it's a test that runs crossbar on Linux (20.04) host by executing a Popen python process: `crossbar start --config coordinator-cfg.yam` and then after that I expected to call `crossbar status` to make sure it's up.
There is a workaround for that by executing the command in a separate process (multiprocessing.Process), or by executing the `crossbar start` with multiprocessing.Process and reading the output, but it would be great if there was a direct way to do that :)
It can also be tested with Linux command `crossbar status | cat` so in this case nothing is printed to console using the pipe.

**Reply by oberstet** [reply]
Thanks for describing your setup/scenario - that is useful to get a picture. However, it would be good if you could just (also) post the full & unmodified output of `crossbar version` verbatim here.

Reason: the command was designed and tuned over time to give complete, thorough version info on all relevant software pieces involved when running Crossbar.io. And many of them might be relevant in the context of this kinda pipes and OS and container and whatnot redirection handling ..

**Reply by chenhayat** [reply]
Sure. that's the output of crossbar version:
 ```
   :::::::::::::::::
          :::::          _____                      __
    :::::   :   :::::   / ___/____ ___   ___  ___  / /  ___ _ ____
    :::::::   :::::::  / /__ / __// _ \ (_-< (_-< / _ \/ _ `// __/
    :::::   :   :::::  \___//_/   \___//___//___//_.__/\_,_//_/
          :::::
    :::::::::::::::::   Crossbar v21.3.1

    Copyright (c) 2013-2024 Crossbar.io Technologies GmbH, licensed under AGPL 3.0.

 Crossbar.io        : 21.3.1
   txaio            : 23.1.1
   Autobahn         : 21.3.1
     UTF8 Validator : autobahn
     XOR Masker     : autobahn
     JSON Codec     : stdlib
     MsgPack Codec  : msgpack-1.0.8
     CBOR Codec     : cbor-1.0.0
     UBJSON Codec   : ubjson-0.16.1
     FlatBuffers    : flatbuffers-latest
   Twisted          : 24.3.0-EPollReactor
   LMDB             : 1.5.1/lmdb-0.9.31
   Python           : 3.11.9/CPython
   PIP              : 24.2
 Frozen executable  : no
 Operating system   : Linux-5.15.0-92-generic-x86_64-with-glibc2.31
 Host machine       : x86_64
 Release key        : RWT906S6dtgOY3NZrHd3uFtXpAPFUc6+bdpADYJRJi4daUaCT5U/0OrO

```

**Reply by oberstet** [reply]
I should probably say: we do have a specific option in Crossbar.io to _activate_ use of `/dev/tty` .. it is pretty deep down buried .. and I also do not understand why it _seems_ to hook in here .. hence I am nagging you for the full version to have a chance to track down whats going on ... anyways, here are some pointers to internal code

https://github.com/crossbario/crossbar/blob/0089c1ef6fbbb87fc7316088a91f1859fa84eeb0/crossbar/shell/repl.py#L164
https://github.com/crossbario/crossbar/blob/master/crossbar/_util.py#L47C5-L47C14

----

fwiw, here is my version info on my local box

```
oberstet@intel-nuci7:~$ cd scm/typedefint/rfminer
oberstet@intel-nuci7:~/scm/typedefint/rfminer$ source .venv/bin/activate
(.venv) oberstet@intel-nuci7:~/scm/typedefint/rfminer$ crossbar version


    :::::::::::::::::
          :::::          _____                 __              _
    :::::   :   :::::   / ___/______  ___ ___ / /  ___ _____  (_)__
    :::::::   :::::::  / /__/ __/ _ \(_-<(_-</ _ \/ _ `/ __/ / / _ \
    :::::   :   :::::  \___/_/  \___/___/___/_.__/\_,_/_/ (_)_/\___/
          :::::
    :::::::::::::::::   Crossbar.io v22.6.1 [19000101-0000000]

    Copyright (c) 2013-2024 Crossbar.io Technologies GmbH. Licensed under EUPLv1.2.

 Crossbar.io        : 22.6.1
   txaio            : 22.2.1
   Autobahn         : 22.7.1
     UTF8 Validator : wsaccel-0.6.2
     XOR Masker     : wsaccel-0.6.2
     JSON Codec     : stdlib
     MsgPack Codec  : msgpack-1.0.4
     CBOR Codec     : cbor2-5.4.3
     UBJSON Codec   : ubjson-0.16.1
     FlatBuffers    : flatbuffers-latest
   Twisted          : 22.8.0-EPollReactor
   LMDB             : 1.3.0/lmdb-0.9.29
   Python           : 3.9.15/CPython
   PIP              : 22.0.4
   NumPy            : 1.23.4
   zLMDB            : 22.6.1
   CFXDB            : 22.6.1
   XBR              : 21.2.1
 Frozen executable  : no
 Operating system   : Linux-5.15.0-122-generic-x86_64-with-glibc2.35
 Host machine       : x86_64
 Release key        : RWRgokUo5rovJw07csyhF67Tw2NOeE7ygo/lxLYK0iF7LVZFxb4O6ti5

(.venv) oberstet@intel-nuci7:~/scm/typedefint/rfminer$ 
```

**Comment by oberstet** [comment]
> Sure. that's the output of crossbar version:

thanks for that! version info: nothing dubious, looks all reasonable.

IOW: no luck wrt to the issue here:(

anyways, sorry, I don't have time digging deep which would be required, or even adding an option to control it which of course would be possible, busy with other stuff ATM.

**Reply by chenhayat** [reply]
Ok Thank you for looking into this!
